spring-cloud-gateway-core/src/main/java/org/springframework/cloud/gateway/filter/factory/rewrite/ModifyResponseBodyGatewayFilterFactory.java

```diff
@@ -185,9 +204,14 @@ ServerHttpResponse decorate(ServerWebExchange exchange) {
 
 					// TODO: flux or mono
 					Mono modifiedBody = clientResponse.bodyToMono(inClass)
-							.flatMap(originalBody -> config.rewriteFunction
+							.flatMap(originalBody -> config.getRewriteFunction()
 									.apply(exchange, originalBody));
 
+					if (config.getEmptyBodySupplier() != null) {
+						modifiedBody = modifiedBody.switchIfEmpty(Mono.defer(() -> (Mono) config.getRewriteFunction()
```

Comment:
I've used *Mono.defer* inside to made avoid supplier calls in the case of non-empty response.
relevant test emptyBodySupplierNotCalledWhenBodyPresent

Reply:
no, It'll be executed only once 
To be sure I wrote a few tests
GatewaySampleApplicationTests.rewriteResponseEmptyBodyToStringWorks
shows that executed  the getRewriteFunction in switchIfEmpty
and
GatewaySampleApplicationTests.emptyBodySupplierNotCalledWhenBodyPresent
shows that code inside switchIfEpty not going to be executed (that's Mono.defer for)
regarding the second test, you may check the route with id *rewrite_response_fail_supplier*
where emptyBodySupplier do throw Exception
```
() -> {
									throw new IllegalArgumentException("this should not happen");
									}
```
and since the  response still OK, that code won't executed

Reply:
Ok, now I get it, didn't see it at first. You're also passing your default response  through the rewrite function.
I'm not sure if this is something we want. If we're providing a default response, why would we want to change it again with the rewrite function?

Reply:
The point of this change, from my perspective, is to have the ability to rewrite response from backend service in one place for all requests (with or without a body) based on the actual response HTTP statuses/headers/body.
**Example:**
Backend request:
HTTP status: 200 and Body: 
will have the modified body:
```
{
     "code": 200,
     "data":{}
}
```
Backend request:
HTTP status: 500 and Body: 
will have the modified body:
```
{
     "code": 500,
     "data":{}
}
```
I can replace
```
private Supplier emptyBodySupplier;
```
to
```
private Function<Exchange,R> gatewayBodySupplier;
```
And call the function with the actual exchange.
But keep simple supplier with static response won't give the flexibility I need.

Reply:
So let's do something like: (Warning: could give compilation issues 😅)
```java
private Function<ServerWebExchange, Publisher<R>> emptyBodyRewriteFunction;

// and

Mono modifiedBody = clientResponse.bodyToMono(inClass)
							.flatMap(originalBody ->
									config.getRewriteFunction().apply(exchange, originalBody));

if (config.getEmptyBodyRewriteFunction() != null) {
	modifiedBody = modifiedBody.switchIfEmpty(Mono.defer(() -> config.getEmptyBodyRewriteFunction()
		.apply(exchange)));
}
```

Then there is enough flexibility for everyone. You could call your "original" rewrite function in your "empty body" rewrite function to support your use case. Should be `null` by default so we don't break existing applications who want to upgrade.

cc @spencergibb

Reply:
Wouldn't it simpler be call `bodyToMono(inClass).switchIfEmpty(if emptybodyrewrite != null ? emptybodyrewrite : Mono.empty())`? then let rewrite function happen?

Reply:
To achieve lazy behavior It should be something like
```
bodyToMono(inClass)
    .switchIfEmpty(Mono.defer(() ->
        config.getEmptyBodyRewriteFunction() != null 
            ? config.getEmptyBodyRewriteFunction() 
           : Mono.empty())
    );
```
I'm not sure is the addition execution Mono.defer even when the body is not empty will make any difference in terms of performance, but seems it is just a different code style which will achieve the same goal.
I used "if" just not to create extra objects, event while they are cheap ones.
@spencergibb should I use if or ternary operator as you asked?

Reply:
what you have above looks easily readable to me.

Reply:
Can you also write this new functionality down in the docs? Especially the part where the "default response" will also be rewritten by the rewrite function.

Reply:
done
I've added EmptyBodyRewriteFunction interface to keep consistent interface with RewriteFunction